loader/src/main/java/net/neoforged/fml/loading/mixin/FMLMixinLaunchPlugin.java

```diff
@@ -88,62 +93,124 @@ public void initializeLaunch(ITransformerLoader transformerLoader, NamedPath[] s
-                                    "fml.modloadingissue.mixin.duplicate_config",
-                                    potential.config(),
-                                    existingInfo,
-                                    modId));
-                        }
-                        if (potential.requiredMods().stream().allMatch(id -> modList.getModFileById(id) != null)) {
-                            Mixins.addConfiguration(potential.config());
-                        } else {
-                            LOGGER.debug("Mixin config {} for mod {} not applied as required mods are missing", potential.config(), modId);
-                        }
-                    }
-                });
-
-        final var configMap = Mixins.getConfigs().stream().collect(
-                Collectors.toMap(Config::getName, Config::getConfig));
-        configAnnotationInfo.forEach((fileName, annotationInfo) -> {
-            if (annotationInfo == null) return;
-
-            final var config = configMap.get(fileName);
+        for (var modFileInfo : modList.getModFiles()) {
+            var modFile = modFileInfo.getFile();
+            for (var mixinConfig : modFile.getMixinConfigs()) {
+                // Validate the mixin version is supported
+                if (!validateMixinBehavior(modFile, mixinConfig)) {
+                    continue;
+                }
+
+                var currentInfo = new AnnotationInfo(modFile, calculateBehaviorVersion(mixinConfig.behaviorVersion()));
+                var existingInfo = configAnnotationInfo.putIfAbsent(mixinConfig.config(), currentInfo);
+                if (existingInfo != null && existingInfo.modFile() != modFile) {
+                    ModLoader.addLoadingIssue(ModLoadingIssue.error(
+                            "fml.modloadingissue.mixin.duplicate_config",
+                            mixinConfig.config(),
+                            existingInfo.modFile)
+                            .withAffectedModFile(modFile));
+                    continue;
+                }
+
+                // Grab the actual mixin config content from the Jar, which also allows us to print a nicer error if it's missing
+                byte[] configContent;
+                try {
+                    configContent = modFile.getContents().readFile(mixinConfig.config());
+                } catch (IOException e) {
+                    configContent = null;
+                }
+                if (configContent == null) {
+                    ModLoader.addLoadingIssue(ModLoadingIssue.error(
+                            "fml.modloadingissue.mixin.missing_config",
+                            mixinConfig.config())
+                            .withAffectedModFile(modFile));
+                    continue;
+                }
+
+                if (mixinConfig.requiredMods().stream().allMatch(id -> modList.getModFileById(id) != null)) {
+                    service.addMixinConfigContent(mixinConfig.config(), configContent);
```

Comment:
As mentioned on discord -- this logic means that a mixin config _must_ come from the mod file declaring it. This is... I mean, frankly, a pretty sensible requirement, right? _But_ it's new compared to the old mixin impl and if we're going to add it we should let multiple mods use the same name the way I described elsewhere.